src/test/crimson/test_config.cc

```diff
@@ -1,27 +1,81 @@
 #include <chrono>
+#include <string>
 #include <numeric>
 #include <seastar/core/app-template.hh>
 #include <seastar/core/sharded.hh>
+#include "common/config_obs.h"
 #include "crimson/common/config_proxy.h"
 
 using Config = ceph::common::ConfigProxy;
+const std::string test_uint_option = "osd_max_pgls";
+
+class ConfigObs : public ceph::internal::md_config_obs_impl<Config> {
+  uint64_t last_change;
+  uint64_t num_changes;
+  Config &local_conf;
+  const char **conf_keys;
+public:
+  ConfigObs(Config::ShardedConfig &conf)
+    : last_change(-1), num_changes(0), local_conf(conf.local()){
+    conf_keys = new const char*[2];
+    conf_keys[0] = test_uint_option.c_str();
+    conf_keys[1] = NULL;
+    local_conf.add_observer(this);
+  }
+  ~ConfigObs() {
+    local_conf.remove_observer(this);
+    delete[] conf_keys;
+  }
+  uint64_t get_last_change() const { return last_change; }
+  uint64_t get_num_changes() const { return num_changes; }
+  const char** get_tracked_conf_keys() const override {
+    return conf_keys;
+  }
+  void handle_conf_change(const Config& conf,
+                          const std::set <std::string> &changes) {
+    if (changes.count(test_uint_option)) {
+      last_change = conf.get_val<uint64_t>(test_uint_option);
+      num_changes += 1;
+    }
+  }
+  seastar::future<> stop() { return seastar::now(); }
+};
+
+seastar::sharded<ConfigObs> sharded_cobs;
 
 static seastar::future<> test_config()
 {
   return ceph::common::sharded_conf().start().then([] {
-    return ceph::common::sharded_conf().invoke_on(0, &Config::start);
-  }).then([] {
-    return ceph::common::sharded_conf().invoke_on_all([](Config& config) {
-      return config.set_val("osd_tracing", "true");
-    });
-  }).then([] {
-    return ceph::common::sharded_conf().invoke_on_all([](Config& config) {
-      if (!config.get_val<bool>("osd_tracing")) {
-        throw std::runtime_error("run osd_tracing");
-      }
+    return ceph::common::sharded_conf().invoke_on(0, &Config::start).then([] {
```

Comment:
@cyx1231st i'd suggest 

- reduce the indent level
- combine the `invoke_on_other()` call with `invoke_on_all()` for readability. yeah, i knew it looks redundant as we `get_val()` and then compare its reval with `get_val()` on the same shard. but it's a lot easier to read this way.
- instead of storing `local_conf` as a member variable always use the helper function `ceph::common::local_conf()`

```c++
  return ceph::common::sharded_conf().start().then([] {
    return ceph::common::sharded_conf().invoke_on(0, &Config::start);
  }).then([] {
    return sharded_cobs.start();
  }).then([] {
    return ceph::common::sharded_conf().invoke_on_all([](Config& config) {
      return config.set_val(test_uint_option,
                            std::to_string(seastar::engine().cpu_id()));
      });
  }).then([] {
    return seastar::do_with(ceph::common::local_conf().get_val<uint64_t>(test_uint_option), [](auto expected) {
      return ceph::common::sharded_conf().invoke_on_all([expected](Config& config) {
        if (expected != config.get_val<uint64_t>(test_uint_option)) {
          throw std::runtime_error("configurations don't match");
        }
        if (expected != sharded_cobs.local().get_last_change()) {
          throw std::runtime_error("last applied changes don't match the latest config");
        }
        if (seastar::smp::count != sharded_cobs.local().get_num_changes()) {
          throw std::runtime_error("num changes don't match actual changes");
        }
      });
    });
  }).finally([] {
    return sharded_cobs.stop();
  }).finally([] {
    return ceph::common::sharded_conf().stop();
  });
```

Reply:
About reducing the indent level, I just thought if something went wrong during or before `sharded_cobs.start()`, `sharded_cobs.stop()` shouldn't be called. And it is similar to `sharded_conf().start()` and `sharded_conf().stop()`. That's why they appear nested, maybe I'm overthinking this issue.

Reply:
i don't think `ConfigProxy` 's constructor throws. if we fail to re-allocate the `_instances` , `std::bad_alloc` is thrown. but probably we can live with this. what do you think?